### ISSUE TITLE
Extending Functionality for Policy and Replay Buffers

### ISSUE THREAD

**Issue by cremebrule** [Denys88/rl_games#63]
Hi,

Had a couple of questions regarding extending the current functionality present in rl_games:

1. What would be the best way to extend one of the algos (say, A2C Continuous) to allow for an external function (like a controller function) to be called after the NN forward pass? For example, normally the forward pass (at inference time) might look like `model_forward()` --> `dist_from_output()` --> `sample_from_dist()`, whereas I'm hoping to inject an external function after the forward pass so that the pipeline would look like `model_forward()` --> `external_postprocessing()` --> `dist_from_output()` --> `sample_from_dist()`, where `external_postprocessing()` would take in the model's outputted values and return the post-processed values (potentially of a different dimension, which would be the "final" action dimension used to generate the sampling distribution; e.g.: conversion from eef commands into joint torques)

2. What would be the best way to include additional information to be stored in the replay buffer (to be used by the above external function)? Ideally, this would be a dict of tensors that is stored along with the normal (s, a, r, s') values for a given env step.

Working with @ViktorM on applications relevant to these features and he thought it might be best if I posted here. Thanks!

**Comment by Denys88** [comment]
Hi, 
 I'll answer till the EOW.

**Comment by Denys88** [comment]
Hi @cremebrule ,
If you want to use different distributions please tale a look at ModelA2CContinuousLogStd and ModelA2CContinuous.
You can take a look how I deal with the masked actions in Discrete model too (class ModelA2C).
First of all please take a look at the play_steps_rnn and play_steps methods. If it is discrete action space and I am using actions mask to remove forbidden actions I need to store all masks from inference for the training. Also ModelA2C uses additional code in case if we have masks:
'''
categorical = CategoricalMasked(logits=logits, masks=action_masks)
'''
and if actions masks are empty I use regular Categorical.
Overall looks like you need to implement your own Model class or extend existing ModelA2CContinuousLogStd.
Whe you do it, you need to register it in the factory in the model_builder.py.

**Comment by cremebrule** [comment]
Hi @Denys88 ,

Thanks, this was helpful!
